README.md
![fah](cdkfah.png)

# CDK-Folding@Home

Run [Folding@Home](https://foldingathome.org/home/) with this prebuild CDK library and never forget to shut off expensive instances.
With built-in TTL for your stack, you just define the lifetime of your instances and the stack will be destroyed afterwards.

### About Folding@Home

Folding@home is a project focused on disease research. The problems F@H is solving require many computer calcul­ations – so you can donate your excess compute power to help their research.
In that sense, you are part of a huge distributed supercomputer!

The architecture follows this AWS [blog post](https://aws.amazon.com/blogs/compute/foldinghome-infectious-disease-research-with-spot-instances/) and leverages g4dn.xlarge Spot EC2 instances to maximise GPU vs price efficiency. 

Running at spot price drastically reduces the price from 0.59$ per hour to around 0.17 in eu-west-1. That's a nice 70% in savings.

![spotsavings](spotpricing.png)

I hope you enjoy this package and together we can make a difference and help the [Folding@Home](https://foldingathome.org/home/) program to succeed in their mission!

### Progress
After deployment, you can go into CloudWatch logs and open the log group *fahlog*.
You can now see the progress every one of your instances makes towards the goal of unfolding proteins.
![wu01fahlog.png](https://cdn.hashnode.com/res/hashnode/image/upload/v1603533266726/2Fcf_74Lt.png)
wu01 represents the work of the GPU and wu00 is the work of the CPU .

Big thanks to [cdk-time-bomb](https://github.com/jmb12686/cdk-time-bomb) for the TTL function and [Ash Belmokadem](https://thecloudboss.nl/) and Vikin Shetty for debugging help!

## How to use it

1. Create a new CDK app

2. Install CDK-Folding@Home in your CDK app

``` Bash
npm install cdkfoldingathome
```

3. Open your *stack*.ts file in ./bin

4. Import cdkfoldingathome

``` Typescript
import * as folding from 'cdkfoldingathome';

```
5. Create folding stack - *important to specify accountId and region*

``` Typescript
const fah = new folding.FoldingathomeStack(app, 'folding', {
    timeToLive: cdk.Duration.hours(5),
    spotPrice: '0.40',
    numberOfInstances: 10,
    env: {
        account: "754950554578",
        region: "eu-west-1"
    }
})
```

6. Ready to build and deploy!

7. Your stack gets destroyed once the TTL is up

8. Double-check the first time to be sure your stack got really deleted 😉

If you have any questions reach out on Github or [Twitter](https://twitter.com/win_bv).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.2.2 (2020-10-26)


test/hello.test.ts
import '@aws-cdk/assert/jest';


test('adds 1 + 2 to equal 3', () => {
  expect(3).toBe(3);
});


src/selfdestruct.ts
import { Rule, Schedule, RuleTargetInput } from '@aws-cdk/aws-events';
import { LambdaFunction } from '@aws-cdk/aws-events-targets';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';

//Construct enables self destruction of the entire CloudFormation Stack for which it is initialized in
//Inspired by https://dev.to/guneysus/developing-a-self-destructing-cloudformation-stack-2ka2

export interface SelfDestructProps {
  readonly timeToLive: cdk.Duration;
}

export class SelfDestruct extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: SelfDestructProps) {
    super(scope, id);

    const selfDestructPolicy: iam.PolicyDocument = new iam.PolicyDocument();

    /** define inline policies */
    const selfDestructPolicyStatements: iam.PolicyStatement[] = [
      //temporarily allow ALL
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['*'],
        resources: ['*'],
      }),
    ];

    selfDestructPolicyStatements.forEach((statement: iam.PolicyStatement) =>
      selfDestructPolicy.addStatements(statement),
    );

    const selfDestructRole = new iam.Role(this, 'SelfDestructLambdaRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AWSLambdaBasicExecutionRole',
        ),
      ],
      inlinePolicies: {
        policy: selfDestructPolicy,
      },
    });

    const selfDestructLambda = new lambda.SingletonFunction(
      this,
      'self-destruct-stack-app',
      {
        uuid: '6024fd94-c050-11ea-b3de-0242ac130004',
        runtime: lambda.Runtime.PYTHON_3_6,
        role: selfDestructRole,
        handler: 'index.handler',
        code: lambda.Code.fromInline(`
import boto3
client = boto3.client('cloudformation')

def handler(event, context):
  print("event['StackName']: "+event['StackName'])
  return client.delete_stack(
    StackName=event['StackName']
  )
            `),
      },
    );

    const rule = new Rule(this, 'SelfDestructRule', {
      schedule: Schedule.rate(props.timeToLive),
    });

    selfDestructLambda.addPermission('allowEventsPermission', {
      action: 'lambda:InvokeFunction',
      principal: new iam.ServicePrincipal('events.amazonaws.com'),
      sourceArn: rule.ruleArn,
    });

    var params = {
      StackName: cdk.Aws.STACK_NAME,
    };

    rule.addTarget(
      new LambdaFunction(selfDestructLambda, {
        event: RuleTargetInput.fromObject(params),
      }),
    );
  }
}

resources/folding.sh
#!/bin/bash -xe
sudo yum -y update
wget https://download.foldingathome.org/releases/public/release/fahclient/centos-6.7-64bit/v7.6/fahclient-7.6.13-1.x86_64.rpm
sudo rpm -i --nodeps fahclient-7.6.13-1.x86_64.rpm
wget https://download.foldingathome.org/releases/public/release/fahcontrol/centos-6.7-64bit/v7.6/fahcontrol-7.6.13-1.noarch.rpm
sudo rpm -i --nodeps fahcontrol-7.6.13-1.noarch.rpm
wget https://download.foldingathome.org/releases/public/release/fahviewer/centos-6.7-64bit/v7.6/fahviewer-7.6.13-1.x86_64.rpm
sudo rpm -i --nodeps fahviewer-7.6.13-1.x86_64.rpm

# Cloud Watch Log
wget https://s3.amazonaws.com/amazoncloudwatch-agent/amazon_linux/amd64/latest/amazon-cloudwatch-agent.rpm
sudo rpm -i --nodeps amazon-cloudwatch-agent.rpm
cat << EOF > /home/ec2-user/cloudwatch_config.json
{
    "agent": {
    "run_as_user": "root"
    },
    "logs": {
            "logs_collected": {
            "files": {
                "collect_list": [
                {
                    "file_path": "/var/lib/fahclient/log.txt",
                    "log_group_name": "fahlog",
                    "log_stream_name": "{instance_id}"
                }
                ]
            }
            }
    }
}
EOF

sudo cp /home/ec2-user/cloudwatch_config.json /opt/aws/amazon-cloudwatch-agent/bin/config.json
sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -m ec2 -c file:/opt/aws/amazon-cloudwatch-agent/bin/config.json -s
# If the gpu v=true it enables the GPU to be utilized
cat << EOF > /home/ec2-user/fah_config.xml
<config>
    <gpu v="true"/>
    <power v="full"/>
    # User information section, please modify accordingly
    <!-- User Information -->
    <team v="Default"/>
    <user v="0"/> 
    <passkey v="PASSKEY_HERE"/>
</config>
EOF
sudo /etc/init.d/FAHClient stop
sudo cp /home/ec2-user/fah_config.xml /etc/fahclient/config.xml
sudo /etc/init.d/FAHClient start

src/index.ts
import * as autoscaling from '@aws-cdk/aws-autoscaling';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import { SelfDestruct } from './selfdestruct';

export interface FoldingAtHomeProps extends cdk.StackProps{
  /**
   * Time to live before the stack gets destroyed
   */
  readonly timeToLive: cdk.Duration;
  /**
   * How many spot instances would you like to deploy?
   * @default 1
   */
  readonly numberOfInstances?: number;
  /**
   * How many spot instances would you like to deploy?
   * @default "0.59"
   */
  readonly spotPrice?: string;
}
export class FoldingathomeStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props: FoldingAtHomeProps) {
    super(scope, id, props);

    const selfDestruct = new SelfDestruct(this, 'selfDestructor', {
      timeToLive: props.timeToLive || cdk.Duration.minutes(1),
    });
    const vpc = new ec2.Vpc(this, 'Vpc', {
      natGateways: 0,
      cidr: '10.0.0.0/16',
      maxAzs: 2,
      enableDnsSupport: true,
      enableDnsHostnames: true,
    });
    vpc.node.addDependency(selfDestruct);

    const instanceRole = new iam.Role(this, 'webinstancerole',
      {
        assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
        managedPolicies: [iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AmazonEC2RoleforSSM')],
      });
    instanceRole.node.addDependency(selfDestruct);

    var bootscript: string;
    bootscript = `#!/bin/bash -xe
sudo yum -y update
wget https://download.foldingathome.org/releases/public/release/fahclient/centos-6.7-64bit/v7.6/fahclient-7.6.13-1.x86_64.rpm
sudo rpm -i --nodeps fahclient-7.6.13-1.x86_64.rpm
wget https://download.foldingathome.org/releases/public/release/fahcontrol/centos-6.7-64bit/v7.6/fahcontrol-7.6.13-1.noarch.rpm
sudo rpm -i --nodeps fahcontrol-7.6.13-1.noarch.rpm
wget https://download.foldingathome.org/releases/public/release/fahviewer/centos-6.7-64bit/v7.6/fahviewer-7.6.13-1.x86_64.rpm
sudo rpm -i --nodeps fahviewer-7.6.13-1.x86_64.rpm

# Cloud Watch Log
wget https://s3.amazonaws.com/amazoncloudwatch-agent/amazon_linux/amd64/latest/amazon-cloudwatch-agent.rpm
sudo rpm -i --nodeps amazon-cloudwatch-agent.rpm
cat << EOF > /home/ec2-user/cloudwatch_config.json
{
    "agent": {
    "run_as_user": "root"
    },
    "logs": {
            "logs_collected": {
            "files": {
                "collect_list": [
                {
                    "file_path": "/var/lib/fahclient/log.txt",
                    "log_group_name": "fahlog",
                    "log_stream_name": "{instance_id}"
                }
                ]
            }
            }
    }
}
EOF

sudo cp /home/ec2-user/cloudwatch_config.json /opt/aws/amazon-cloudwatch-agent/bin/config.json
sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -m ec2 -c file:/opt/aws/amazon-cloudwatch-agent/bin/config.json -s
# If the gpu v=true it enables the GPU to be utilized
cat << EOF > /home/ec2-user/fah_config.xml
<config>
    <gpu v="true"/>
    <power v="full"/>
    # User information section, please modify accordingly
    <!-- User Information -->
    <team v="Default"/>
    <user v="0"/> 
    <passkey v="PASSKEY_HERE"/>
</config>
EOF
sudo /etc/init.d/FAHClient stop
sudo cp /home/ec2-user/fah_config.xml /etc/fahclient/config.xml
sudo /etc/init.d/FAHClient start`;

    const asg = new autoscaling.AutoScalingGroup(this, 'ASG', {
      vpc: vpc,
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.GRAPHICS4_NVME_DRIVE_HIGH_PERFORMANCE, ec2.InstanceSize.XLARGE),
      machineImage: ec2.MachineImage.lookup({
        name: 'Deep Learning AMI (Amazon Linux) Version 27.0',
        owners: ['898082745236'],
      }),
      desiredCapacity: props.numberOfInstances,
      maxCapacity: props.numberOfInstances,
      role: instanceRole,
      vpcSubnets: { subnetType: ec2.SubnetType.PUBLIC },
      spotPrice: props.spotPrice,
    });
    asg.addUserData(bootscript);
    asg.node.addDependency(selfDestruct);

  }
}

